### DISCUSSION TITLE
How to understand the output shape of my custom trained model?

### DISCUSSION THREAD

**Discussion by JohannesBauer97** [ultralytics/yolov5#9831] [Q&A] [answered]
Hey everyone,

I trained a model to identify **41** different objects and exported it as MLModel for Apple.
Currently I'm discovering the model in swift and after predicting one image it results in `MultiArray : Float32 1 × 25500 × 46 array`

After reading https://github.com/ultralytics/yolov5/discussions/7774 my understanding of this output is:
1. Dimension = **Batch size**, which is always one
2. Dimension = **Predictions?** At least this is what I take from here: https://github.com/ultralytics/yolov5/discussions/4054#discussioncomment-1020704
3. Dimension = **4 Coordinates (xywh), 41 Classes, 1 Confidence**

So I'm interested in the 3rd dimension and iterated through it using this peace of swift code:
````
for i in 0...45 {
      let key = [0,25499,i] as [NSNumber]
      let val = multiArray[key]
      print(val)
}

/* Output:
595.0998
617.9398
982.113
626.5577
6.238096e-05
0.01111226
0.01590065
0.01978323
0.01647663
0.02073891
0.01734269
0.02134776
(.. and more)
*/
````

_What are these floats telling me now?_
This is my current guess:

````
595.0998 // x - coord
617.9398 // x - coord
982.113 // full width
626.5577 // full height
6.238096e-05 // overall confidence
0.01111226 // class 1 prediction confidence
0.01590065 // class 2 prediction confidence
0.01978323 // ...
0.01647663
0.02073891
0.01734269
0.02134776
(.. and more)
````

### tl;dr: Am I right with the following statement?
After I get a result from the model, I need to sort the 2nd dimension arrays by their 5th value (overall confidence) to get the top predictions and afterwards access the 3rd dimension to find the highest class confidence and it's according coordinates.

**Comment by glenn-jocher** [comment, marked as answer]
@JohannesBauer97 yes that seems correct. Recommend https://hub.ultralytics.com for training and export of YOLOv5 CoreML models, it provides automatic NMS pipelining and allows for Previewing the model in Xcode:

<img width="1290" alt="Screenshot 2022-10-17 at 19 04 31" src="https://user-images.githubusercontent.com/26833433/196239529-05440358-4a49-4f20-a20a-e46f2d7ca986.png">

**Reply by JohannesBauer97** [reply]
Ok, that answer took less than 5 mins, thank you!
I'll definitely check out the hub.
